statsd/statsd.go

```diff
@@ -489,15 +489,23 @@ func (c *Client) SimpleServiceCheck(name string, status ServiceCheckStatus) erro
 	return c.ServiceCheck(sc)
 }
 
+func (c *Client) shutdown() {
+	c.Lock()
+	defer c.Unlock()
+	select {
+	case <-c.stop:
+		return
+	default:
+	}
+	close(c.stop)
+}
+
 // Close the client connection.
 func (c *Client) Close() error {
 	if c == nil {
 		return ErrNoClient
 	}
-	select {
-	case c.stop <- struct{}{}:
-	default:
-	}
+	c.shutdown()
```

Comment:
I'd like to recommend a simpler approach. Where the `watch` and `telemetry` goroutines are started, add:
```
c.wg.Add(1)
go func() {
    defer c.wg.Done()
    c.watch()
}()
c.wg.Add(1)
go func() {
    defer c.wg.Done()
    c.telemetry()
}()
```
And in this function (`Close`) just do:
```
close(c.stop)
c.wg.Wait()
```

Reply:
I don't know if we need the wait group, but I can add it. 

I wrote and used the `shutdown()` function rather than just `close(c.stop)` to prevent a panic if `Close()` was called more than once. 

I looked closer and realize that it will panic anyway if called more than once in `c.sender.close()`, so removing `shutdown()` and just calling `close(c.stop)` is sufficient.

Reply:
The definition of a graceful shutdown is to wait for all goroutines to exit. If you exit the process, or close the statsd client but don't wait for the goroutines to exit, that's also a leak.

Closing that channel doesn't guarantee that the goroutine exits and finishes before the `Close` call returns. Technically, it is definitely possible for a `Close` call to finish and the statsd client might not be closed because the `select` loop in any of those goroutines can always trigger the other case before the stop channel.

Reply:
👍 I'll add the waitgroup stuff to ensure the routines have shut down before returning.